env/io_posix.cc

```diff
@@ -1543,8 +1543,13 @@ PosixDirectory::PosixDirectory(int fd) : fd_(fd) {
 #ifdef OS_LINUX
   struct statfs buf;
   int ret = fstatfs(fd, &buf);
+#if UINTPTR_MAX == 0xffffffff
+  is_btrfs_ =
+      (ret == 0 && static_cast<uint32_t>(buf.f_type) == BTRFS_SUPER_MAGIC);
+#else
   is_btrfs_ = (ret == 0 && buf.f_type == BTRFS_SUPER_MAGIC);
```

Comment:
Conditional compilation is ugly and error-prone, especially since we don't yet have 32-bit coverage in CI. Does this work? `buf.f_type == decltype(buf.f_type){BTRFS_SUPER_MAGIC}`

Also there's a similar comparison to ZFS_SUPER_MAGIC above in the same file.

Reply:
@pdillinger We have been working to add 32bit compilation to CI, but that is a more tricky task which we have almost completed. In the meantime however this is needed for 32bit RocksJava binaries that we release (cc @riversand963). 

I will try the `decltype` option you have suggested against all the release platforms and come back to you...

Reply:
@pdillinger I am afraid your suggestion to use `decltype` has a similar problem to the original compilation issue, i.e.:

```
env/io_posix.cc: In constructor 'rocksdb::PosixDirectory::PosixDirectory(int)':
env/io_posix.cc:1546:80: error: narrowing conversion of '2435016766u' from 'unsigned int' to 'int' inside { } [-Werror=narrowing]
   is_btrfs_ = (ret == 0 && buf.f_type == decltype(buf.f_type){BTRFS_SUPER_MAGIC});
                                                                                ^
cc1plus: all warnings being treated as errors
make[1]: *** [jls/env/io_posix.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory `/rocksdb-local-build'
```

Reply:
@pdillinger Ah ha! Thanks that seems to have done the trick - I have pushed an additional commit.